### PR TITLE
Migration of Advanced Parameters > Team > Profiles page.

### PR DESCRIPTION
**Pull request by rokaszygmantas**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | This PR migrates the profiles list page (Advanced Parameters > Team > Profiles).
| Type?         | improvement
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/issues/10505
| How to test?  | Compare to the legacy controller. This PR migrates only the listing page of Profiles. URL: `admin-dev/index.php/configure/advanced/profiles/`. **You have to rebuild the assets before testing.** Superadmin specific rules will be done in another PR (see https://github.com/PrestaShop/PrestaShop/issues/10505)

- [x] Create grid
- [x] Implement grid actions
- [x] Specific logic on grid bulk/row actions

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/10783)

### CODE REVIEW

File: src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfilesController.php [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+     *
+     * @return RedirectResponse
+     */
+    public function searchAction(Request $request)
+    {
+        $definitionFactory = $this->get('prestashop.core.grid.definition.factory.profiles');
+        $definitionFactory = $definitionFactory->getDefinition();
+
+        $gridFilterFormFactory = $this->get('prestashop.core.grid.filter.form_factory');
+        $searchParametersForm = $gridFilterFormFactory->create($definitionFactory);
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfilesController.php:89]: should be createForm($definitionFactory), I've forgotten to update it :/

Comment (sarjon) [reply src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfilesController.php:89]: :+1: for `createForm` and not `getForm`. :p

Comment (PierreRambaud) [reply src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfilesController.php:89]: Why redundant `Form`, a form factory can only create form... `create` is the right name

Comment (sarjon) [reply src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfilesController.php:89]: @PierreRambaud where were you when we had discussion with @mickaelandrieu about that for other services. -_- i needed your backup. :p

Comment (PierreRambaud) [reply src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfilesController.php:89]: Didn't see the discussion... Too many things to do :'(

Comment (sarjon) [reply src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfilesController.php:89]: in short, it was about `GridDefinitionFactoryInterface::create()` => `GridDefinitionFactoryInterface::getDefinition()`

Comment (PierreRambaud) [reply src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfilesController.php:89]: Always call me when you want someone to do mess with naming :smile:

Comment (sarjon) [reply src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfilesController.php:89]: we may still fix that untill 1.7.5 release if you get @mickaelandrieu on your side. :smile:

Comment (PierreRambaud) [reply src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfilesController.php:89]: @matks wdyt too?

Comment (matks) [reply src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfilesController.php:89]: I'm sorry, I'm going to disappoint all of you ^^ : I think both names are right. One is more redundant, but both names provide the required information ("what is this function doing ?") very accurately so I do not care.

File: src/Core/Grid/Definition/Factory/ProfilesGridDefinitionFactory.php [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        $this->resetActionUrl = $resetActionUrl;
+        $this->redirectionUrl = $redirectionUrl;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getId()
+    {
+        return 'profiles';
```
Comment (tomas862) [inline src/Core/Grid/Definition/Factory/ProfilesGridDefinitionFactory.php:86]: All new grid ids starts with  upper-case 

```suggestion
        return 'Profiles';
```

File: src/Core/Grid/Query/ProfilesQueryBuilder.php [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+     * @param Connection $connection
+     * @param string $dbPrefix
+     * @param DoctrineSearchCriteriaApplicatorInterface $searchCriteriaApplicator
+     * @param int $languageId
+     */
+    public function __construct(
+        Connection $connection,
+        $dbPrefix,
+        DoctrineSearchCriteriaApplicatorInterface $searchCriteriaApplicator,
+        $languageId
```
Comment (tomas862) [inline src/Core/Grid/Query/ProfilesQueryBuilder.php]: Maybe a better naming of this id would be $contextLanguageId ? :thinking: 
```suggestion
        $contextLanguageId
```

File: src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfilesController.php [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        }
+
+        return $this->redirectToRoute('admin_profiles', ['filters' => $filters]);
+    }
+
+    /**
+     * Shows profile edit form.
+     *
+     * @AdminSecurity(
+     *     "is_granted('update', request.get('_legacy_controller'))",
```
Comment (tomas862) [inline src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfilesController.php:105]: Provide redirectRoute parameter

```suggestion
     *     "is_granted('update', request.get('_legacy_controller'))",
     *     redirectRoute="admin_profiles",
```

File: src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfilesController.php [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+        ]);
+
+        return $this->redirect($legacyLink);
+    }
+
+    /**
+     * Delete a profile.
+     *
+     * @AdminSecurity(
+     *     "is_granted(['delete'], request.get('_legacy_controller')~'_')",
```
Comment (tomas862) [inline src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfilesController.php]:
```suggestion
     *     "is_granted('delete', request.get('_legacy_controller')~'_')",
```

File: src/PrestaShopBundle/Resources/config/routing/admin/configure/advanced_parameters/profiles.yml [hunk 6]
```diff
@@ -0,0 +1,38 @@
+admin_profiles:
```
Comment (tomas862) [inline src/PrestaShopBundle/Resources/config/routing/admin/configure/advanced_parameters/profiles.yml]: Check naming convention :+1: 
```suggestion
admin_profiles_index:
```

File: src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfilesController.php [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+ * "Configure > Advanced parameters > Team > Profiles" page.
+ */
+class ProfilesController extends FrameworkBundleAdminController
+{
+    /**
+     * Show profiles listing page.
+     *
+     * @AdminSecurity("is_granted('read', request.get('_legacy_controller'))")
+     *
+     * @Template("@PrestaShop/Admin/Configure/AdvancedParameters/Profiles/profiles.html.twig")
```
Comment (tomas862) [inline src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfilesController.php]: Check naming convention :+1: 

```suggestion
     * @Template("@PrestaShop/Admin/Configure/AdvancedParameters/Profiles/index.html.twig")
```

File: src/Core/Grid/Definition/Factory/ProfilesGridDefinitionFactory.php (hunk 2 above)
Comment (mickaelandrieu) [reply src/Core/Grid/Definition/Factory/ProfilesGridDefinitionFactory.php:86]: is it a convention? If yes, it needs to be documented.

File: src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfilesController.php (hunk 1 above)
Comment (mickaelandrieu) [reply src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfilesController.php:89]: I think it's too late to change the name, but if we can I stand for `getForm()` for consistency with all others functions in the Grid component.

**Comment by rokaszygmantas** [comment]
Hi @matks,

This listing page migrated, except there is one thing that is different compared to legacy controller, and it can be seen in these screenshots:
Legacy: https://prnt.sc/lxvy86
Migrated: https://prnt.sc/lxvy2v
The bulk actions are rendered for all rows and the accessibility checking is not implemented for bulk actions in the grid component yet. I've talked with @sarjon about this and we came to a conclusion that it would be better to open another PR with this improvement to grid component, because it's a global change needed for many grid implementations. Wdyt?

Btw the backend accessibility checking is implemented, so you cannot delete the super admin profile, it's just that the checkbox is displayed visually.

**Comment by matks** [comment]
@rokaszygmantas Great 👍 

Yes it's nice to split deliverables into several PRs. I added this item to [the issue todlist](https://github.com/PrestaShop/PrestaShop/issues/10505#issue-361803488) so we dont forget.

Now I've got to review 😄

File: src/Core/Grid/Query/ProfilesQueryBuilder.php [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+    {
+        $qb = $this->connection
+            ->createQueryBuilder()
+            ->from($this->dbPrefix . 'profile', 'p')
+            ->innerJoin('p', $this->dbPrefix . 'profile_lang', 'pl', 'p.id_profile = pl.id_profile')
+            ->andWhere('pl.id_lang = :language')
+            ->setParameter('language', $this->languageId)
+        ;
+
+        foreach ($filters as $name => $value) {
```
Comment (matks) [inline src/Core/Grid/Query/ProfilesQueryBuilder.php:116]: Please restrict the possible `$filters` value. We know this reduces the reusability of this class, but security is more important
```
    $allowedFilters = [ ... ];

     foreach ($filters as $name => $value) {
            if (!in_array($filterName, $allowedFilters, true)) {
                continue;
            }
```

File: src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfilesController.php [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+/**
+ * Class ProfilesController is responsible for displaying the
+ * "Configure > Advanced parameters > Team > Profiles" page.
+ */
+class ProfilesController extends FrameworkBundleAdminController
+{
+    /**
+     * Show profiles listing page.
+     *
+     * @AdminSecurity("is_granted('read', request.get('_legacy_controller'))")
```
Comment (matks) [inline src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfilesController.php:51]: We have an ongoing debate about whether we should add `create`, `update` and `delete` rights here
https://github.com/PrestaShop/PrestaShop/issues/12109

**Review by matks — COMMENTED** [review]
Almost nothing more to add than Tomas and Mickael 😄

**Comment by matks** [comment]
@rokaszygmantas (finally) reviewed ! sorry for the wait ...

File: src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfilesController.php (hunk 9 above)
Comment (matks) [reply src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfilesController.php:51]: We decided that `read` right should be the only one giving access to listing page. So this code is valid 😄 👍

**Comment by rokaszygmantas** [comment]
@matks rebased, everything is done from my side on this PR. :slightly_smiling_face:

**Comment by matks** [comment]
@rokaszygmantas I'm sorry I dont see the fix for https://github.com/PrestaShop/PrestaShop/pull/10783/files#r247077480 🤔

**Comment by rokaszygmantas** [comment]
@matks sorry, must have missed it! I'll fix it and review other comments once more :)

**Comment by rokaszygmantas** [comment]
@matks fixed it :)

**Review by matks — DISMISSED** [review]

**Comment by marionf** [comment]
@TristanLDD 

Is it ok for you ?

capture d ecran_942

**Comment by TristanLDD** [comment]
Hello,

2 little fixes on this page: 

- Add placeholders in search fields 

-Reduce the ID search field to 150px width 

Thank you !

**Comment by rokaszygmantas** [comment]
Hi @marionf, I've fixed the PR according to feedback from @TristanLDD (added placeholders and fixed ID input width). :slightly_smiling_face: 

@matks - I've added a new form type for grid search by ID (`\PrestaShopBundle\Form\Admin\Type\SearchByIdType`), which will have reduced width automatically, so it can be reused across other grids for consistency, wdyt?

**Comment by matks** [comment]
@TristanLDD Screenshot after Rokas changes:
capture d ecran 2019-01-24 a 17 56 15

**Comment by matks** [comment]
@rokaszygmantas Mmmm ... isn't it a bit overkill to add a new FilterType only for CSS display ? 😅 

Maybe we can make it rather an option of current `TextType`, something like
```
->setTypeOptions([
                    'size' => 'small' // we could choose either 'small', 'big' or 'full', full is to use the whole space
                ])
```

**Comment by rokaszygmantas** [comment]
oh well, I liked the idea of having a separate type :smile: but I can change it to a size option for `TextType`, not a big deal :slightly_smiling_face:

**Comment by matks** [comment]
@rokaszygmantas If the new Type defines a different backend behavior, why not. But here its only purpose is to be used as a marker for specific CSS. In theory, the Grid Component is not display-related (it's rather: we have a Grid component and we built on top of it a templating system using Twig).

Moreover, this "size" option will become useful if tomorrow we get to migrate a grid with a "ISO Code" column for example (so max characters length is 3) and Design Team asks us to make it small as well. You can't use a SearchById Type for a SearchByIsoCode field, right 😉 ?

**Comment by rokaszygmantas** [comment]
@matks nice explanation - makes sense!

**Comment by TristanLDD** [comment]
@matks Thank you, design is Ok for me

**Comment by rokaszygmantas** [comment]
Hi @matks, 

I've removed the `SearchById` form type now and introduced new `size` option for `TextType` types. 

I didn't implement other size options besides `small`, they can be easily added when needed. :slightly_smiling_face:

**Review by matks — APPROVED** [review]

**Comment by matks** [comment]
I think we dont need to bother QA to review the UX feedback 😄

**Comment by matks** [comment]
Thank you @rokaszygmantas !

**Pull request merged**